tests/PactBuilder/JsonPatternBuilderTest.php

```diff
@@ -17,87 +18,231 @@ public function testItBuildsMessage(): void
-                'type' => $builder->like('value'),
-                'some' => $builder->term('value', '/.*/'),
-            ]),
-            'some' => $builder->term('value', '/.*/'),
-            'another' => [
-                'deep' => $builder->term('value', '/.*/'),
-            ],
-            'array' => [
-                $builder->term('value', '/.*/'),
-            ],
+            'some' => 'value',
         ]);
 
         $body = $builder->build();
 
         Assert::assertSame(
-            '{"only":"value","type":"value","deep_only":{"some":"value"},"deep_type":{"type":"value","some":"value"},"some":"value","another":{"deep":"value"},"array":["value"]}',
+            '{"some":"value"}',
             $body->content()
         );
 
-        Assert::assertSame([
-            '$.type' => [
-                'combine' => 'AND',
+        Assert::assertSame([], $body->matchingRules()->toArray());
+    }
+
+    public function testItIsImmutable(): void
+    {
+        $builder = new JsonPatternBuilder();
+
+        Assert::assertNotSame($builder, $builder->withPattern([]));
+    }
+
+    public function testItGeneratesWithSimplePattern(): void
+    {
+        $builder = new JsonPatternBuilder();
+        $builder = $builder->withPattern([
+            'some' => [
+                'deep' => 'value',
+            ],
+        ]);
+
+        $body = $builder->build();
+
+        Assert::assertSame(\json_encode([
+            'some' => [
+                'deep' => 'value',
+            ],
+        ], \JSON_THROW_ON_ERROR), $body->content());
+
+        Assert::assertSame(MatchingRules::fromArray([
+        ])->toArray(), $body->matchingRules()->toArray());
+    }
+
+    public function testItGeneratesWithTypePattern(): void
+    {
+        $builder = new JsonPatternBuilder();
+        $builder = $builder->withPattern([
+            'some' => $builder->like('value'),
+        ]);
+
+        $body = $builder->build();
+
+        Assert::assertSame(\json_encode([
+            'some' => 'value',
+        ], \JSON_THROW_ON_ERROR), $body->content());
+
+        Assert::assertSame(MatchingRules::fromArray([
+            '$.some' => [
                 'matchers' => [
                     ['type' => 'type'],
                 ],
             ],
-            '$.deep_only' => [
-                'combine' => 'AND',
+        ])->toArray(), $body->matchingRules()->toArray());
+    }
+
+    public function testItGeneratesWithExpressionPattern(): void
+    {
+        $builder = new JsonPatternBuilder();
+        $builder = $builder->withPattern([
+            'some' => $builder->term('value', '/.*/'),
+        ]);
+
+        $body = $builder->build();
+
+        Assert::assertSame(\json_encode([
+            'some' => 'value',
+        ], \JSON_THROW_ON_ERROR), $body->content());
+
+        Assert::assertSame(MatchingRules::fromArray([
+            '$.some' => [
+                'matchers' => [
+                    ['type' => 'regex', 'pattern' => '/.*/'],
```

Comment:
The regex delimiters shouldn't appear here (this does not refer specifically to this unit test). The specification examples don't have delimiters for their regexes.